### ISSUE TITLE
[Feature Request] Linux version?

### ISSUE THREAD

**Issue by 0hStormy** [squiggythings/WaveTracker#6]
**Is your feature request related to a problem? Please describe.**
I would like to see a Linux version of WaveTracker in the future. Currently works with [wine](https://www.winehq.org/) if you install .Net 8

**Describe the solution you'd like**
Any Linux binary, preferably something universal. Ex: flatpak, AppImage, Distro package.

**Describe alternatives you've considered**
Use [wine](https://www.winehq.org/). But wine can have issues.

**Additional context**
.Net is on Linux and is in most Distro's repositories.

**Comment by 0hStormy** [comment]
I'll see if I can compile it myself rq

**Comment by SolarLune** [comment]
> I'll see if I can compile it myself rq Update: You will have to change the build target to be multi-platform in Visual Studio

Were you able to compile it successfully for Linux? On my machine, running through Wine just gives a black screen.

**Comment by kijemotan** [comment]
can confirm wine just gives a black screen, though the UI itself is still usable (e.g. the open file dialog)

**Comment by Speykious** [comment]
I just stumbled on this project and it looks amazing! When I saw the UI I got excited, because it looked custom-made, so I thought maybe it was cross-platform. Seems not unfortunately.

That said, I've been looking around in the codebase, and it seems like the only Windows-specific things being used are the `MonoGame.Framework.WindowsDX` runtime, and also Windows Forms seemingly just for dialog windows.
The files that import `System.Windows.Forms` are just a handful:
```cs
App.cs                      // There's a closing form for asking to save changes
Forms/EnterText.cs          // The whole class is a form to enter some text
Forms/EnterText.Designer.cs // Generated UI code, with labels, a text box, and OK + Cancel buttons
Tracker/SaveLoad.cs         // Uses OpenFileDialog and SaveFileDialog for various import/export operations 
UI/InstrumentBank.cs        // Has a dialog to rename the instrument
UI/NumberBox.cs             // Starts an EnterText dialog when element is double clicked
UI/NumberBoxDecimal.cs      // Starts an EnterText dialog when element is double clicked
UI/TextBox.cs               // Starts an EnterText dialog when element is clicked and editable
```

Since Monogame has a cross-platform [DesktopGL](https://www.nuget.org/packages/MonoGame.Framework.DesktopGL) runtime, I think it's definitely possible to make a Linux port or even just to make this cross-platform.
The file dialogs could use [NativeFileDialogs.NET](https://github.com/Speykious/NativeFileDialogs.NET), but I don't know for `EnterText`. Maybe a custom-made form, since pretty much the entire UI seems to be made with Monogame.

**Comment by tgpholly** [comment]
I had a look at getting this running natively myself and have had some success the usage of WinForms isn't too bad so that can be worked around, however a big sticking issue is that NAudio doesn't support Linux very well if at all. Commenting out some usages of NAudio is enough to get the app to launch, albeit with no working midi or audio as both rely on Windows libraries.
![image](https://github.com/user-attachments/assets/ae0604a8-0c16-4e0c-97a6-959fcbea8c52)

**Comment by Speykious** [comment]
Oh. Well that sucks. :(

The only other audio library I know for C# is BASS, which osu!lazer uses. It's proprietary, but it's free for non-commercial use. The commercial license is quite expensive though...

**Comment by tgpholly** [comment]
BASS is good, but yeah proprietary, and would require redoing a lot, not really worth.
I feel like it would be more worthwhile contributing upstream to get better Linux support into NAudio.
Regardless of that some things *will* need to change here as well due to WaveTracker it's self using Wasapi from NAudio, which of course will never be available outside of Windows or WINE.

**Comment by Speykious** [comment]
I managed to make it show the UI as well. I had to basically comment out all the Windows Forms and NAudio/Wasapi code...

![image](https://github.com/user-attachments/assets/ccb361db-f2ba-4cf8-98c6-9ce39109123f)

**Comment by squiggythings** [comment]
This is definitely something high on my priority list. Yes- really the only major things holding this back from being cross platform is the WASAPI audio (which I hope would be easy to swap out with any other audio API, all im doing is feeding it a buffer of samples... but I haven't explored this yet) and the Text/File dialogs.
I'd really like in the future to have the custom-made TextBox and NumberBox to accept typing input, so that it matches the aesthetic of the other UI elements, provides better UX since you'd be able to type right in the box. Plus they'd be able to work cross-platform.
I also might not be opposed to building custom file dialogs, after all I kind of did that with the sample browser, but I think using the platform's native dialog is the better move.

In the meantime, was anyone able to get it running in WINE? and if so what steps needed to be taken? If I could provide a WINE-compatible build that would give me a bit of peace of mind that Linux and MacOS users aren't _completely_ excluded.

**Comment by 0hStormy** [comment]
> This is definitely something high on my priority list. Yes- really the only major things holding this back from being cross platform is the WASAPI audio (which I hope would be easy to swap out with any other audio API, all im doing is feeding it a buffer of samples... but I haven't explored this yet) and the Text/File dialogs. I'd really like in the future to have the custom-made TextBox and NumberBox to accept typing input, so that it matches the aesthetic of the other UI elements, provides better UX since you'd be able to type right in the box. Plus they'd be able to work cross-platform. I also might not be opposed to building custom file dialogs, after all I kind of did that with the sample browser, but I think using the platform's native dialog is the better move.
> 
> In the meantime, was anyone able to get it running in WINE? and if so what steps needed to be taken? If I could provide a WINE-compatible build that would give me a bit of peace of mind that Linux and MacOS users aren't _completely_ excluded.

I was able to get it running in Wine with the 64bit version from itch.io, I'm not sure about other distros but Fedora works.

**Comment by Speykious** [comment]
> I'd really like in the future to have the custom-made TextBox and NumberBox to accept typing input, so that it matches the aesthetic of the other UI elements, provides better UX since you'd be able to type right in the box. Plus they'd be able to work cross-platform.
> I also might not be opposed to building custom file dialogs, after all I kind of did that with the sample browser, but I think using the platform's native dialog is the better move.

...Well your comment falls perfectly at the right time, because I just finished replacing the text and file dialogs on my fork, and I was gonna show that xD

For the Windows Forms popups, I replaced it with a `Dialog` since that class already existed as a custom UI. The text input is very very very simple, it can just add a character to the end or remove at the end and validate by pressing enter, but it can become much better (although it would indeed be even better if the input was just happening directly inside the textbox instead of a dialog).
I'm not sure about the text dialogs, because since they work through callbacks I had to slightly change the order of updates, making sure that it checked `ValueWasChangedInternally` before calling `.Update()` or overwriting the value. Also I might've missed some input boxes.

The file dialog extension filters are wrong (reversed, probably) but apart from that it seems ok.
Also I don't show it at the end but I indeed have a `thnig` file saved at the right place.

https://github.com/user-attachments/assets/06576152-5e70-40e5-b77c-90e06968e956

**Comment by SolarLune** [comment]
> > This is definitely something high on my priority list. Yes- really the only major things holding this back from being cross platform is the WASAPI audio (which I hope would be easy to swap out with any other audio API, all im doing is feeding it a buffer of samples... but I haven't explored this yet) and the Text/File dialogs. I'd really like in the future to have the custom-made TextBox and NumberBox to accept typing input, so that it matches the aesthetic of the other UI elements, provides better UX since you'd be able to type right in the box. Plus they'd be able to work cross-platform. I also might not be opposed to building custom file dialogs, after all I kind of did that with the sample browser, but I think using the platform's native dialog is the better move.
> > In the meantime, was anyone able to get it running in WINE? and if so what steps needed to be taken? If I could provide a WINE-compatible build that would give me a bit of peace of mind that Linux and MacOS users aren't _completely_ excluded.
> 
> I was able to get it running in Wine with the 64bit version from itch.io, I'm not sure about other distros but Fedora works.

Are you actually able to use it under Wine, or is it just a black screen? If it does work, how did you resolve the black screen issue?

**Comment by Spyro24** [comment]
For me, it just crashes. It just doesn't show anything except an error in the terminal. (There should finally be a Linux version.) Here is the error:

wine WaveTracker.exe 
wine: Call from 000000007BC2ED58 to unimplemented function KERNEL32.dll.GetProcessGroupAffinity, aborting
wine: Unhandled page fault on read access to 000000000000000C at address 0000000000990EF8 (thread 0168), starting debugger...
Unhandled exception: page fault on read access to 0x0000000c in 64-bit code (0x0
000000000990ef8).
Register dump:
 rip:0000000000990ef8 rsp:000000000029d250 rbp:000000000029f050 eflags:00010206 
(  R- --  I   - -P- )
 rax:00000000000db340 rbx:000000000029d2a0 rcx:0000000000000130 rdx:000000000000
0000
 rsi:000000000029ef20 rdi:000000000029dca0  r8:00000000000003c9  r9:000000001993
0520 r10:0000000000000005
 r11:000000000029d5c0 r12:0000000000000000 r13:000000000029d3c8 r14:000000000029
fea0 r15:000000000029de08
Stack dump:
0x000000000029d250:  000000000029de08 000000000029fea0
0x000000000029d260:  0000000000a5f970 000000007b070687
0x000000000029d270:  000000000029f0f0 0000000000b52793
0x000000000029d280:  0000000000000004 000000000029d550
0x000000000029d290:  000000000029ef20 0000000000d89ef0
0x000000000029d2a0:  0000000000000000 0000000000a5e490
0x000000000029d2b0:  0000000000000000 0000000000a80257
0x000000000029d2c0:  0000000000000000 0000000000d89ef0
0x000000000029d2d0:  000000000029dca0 0000000000a5e3d1
0x000000000029d2e0:  000000000029d3c8 000000000029ef20
0x000000000029d2f0:  000000000029dca0 000000000029d550
0x000000000029d300:  000000000029ea20 0000000000a60980
Backtrace:
=>0 0x0000000000990ef8 EntryPoint+0xffffffffffffffff() in coreclr (0x00000000002
9f050)
  1 0x0000000000b52793 EntryPoint+0xf5c42() in coreclr (0x000000000029f050)
  2 0x0000000000a80257 EntryPoint+0x23706() in coreclr (0x000000000029f050)
  3 0x0000000000a60980 EntryPoint+0x3e2f() in coreclr (0x000000000029d550)
  4 0x0000000000a5fa85 EntryPoint+0x2f34() in coreclr (0x000000000029d550)
  5 0x000000007bc525a5 EntryPoint+0xffffffffffffffff() in ntdll (0x000000000029d
550)
  6 0x00000000009f9e85 EntryPoint+0xffffffffffffffff() in coreclr (0x00000000002
9f3d1)
0x0000000000990ef8 EntryPoint+0xffffffffffffffff in coreclr: movl       0x000000
000000000c(%rdx),%eax
Modules:
Module  Address                                 Debug info      Name (36 modules
)
PE                890000-          8f4000       Deferred        hostpolicy
PE                900000-          de9000       Export          coreclr
PE              7b000000-        7b3fd000       Deferred        kernelbase
PE              7b600000-        7b969000       Deferred        kernel32
PE              7bc00000-        7bf3c000       Dwarf           ntdll
PE             140000000-       140045000       Deferred        wavetracker
PE             180000000-       180059000       Deferred        hostfxr
PE             1c8db0000-       1c9097000       Deferred        msvcrt
PE             1cd860000-       1cd86f000       Deferred        api-ms-win-crt-u
tility-l1-1-0
PE             1d97a0000-       1d97ae000       Deferred        api-ms-win-core-
fibers-l1-1-1
PE             21a7e0000-       21a9ce000       Deferred        setupapi
PE             231ae0000-       231d60000       Deferred        rpcrt4
PE             23d820000-       23dfc2000       Deferred        user32
PE             241850000-       24185e000       Deferred        api-ms-win-crt-e
nvironment-l1-1-0
PE             262250000-       262260000       Deferred        api-ms-win-crt-r
untime-l1-1-0
PE             26b4c0000-       26bbb9000       Deferred        gdi32
PE             2739c0000-       273ec9000       Deferred        oleaut32
PE             28ba60000-       28ba6e000       Deferred        api-ms-win-crt-t
ime-l1-1-0
PE             2e3540000-       2e364f000       Deferred        shlwapi
PE             2e8f10000-       2e948f000       Deferred        ole32
PE             2f1fa0000-       2f1fbe000       Deferred        version
PE             30a2c0000-       30a2d0000       Deferred        api-ms-win-crt-s
tdio-l1-1-0
PE             30c980000-       30c98f000       Deferred        api-ms-win-core-
synch-l1-2-0
PE             3126f0000-       312744000       Deferred        shcore
PE             327020000-       3271a4000       Deferred        combase
PE             32a700000-       32a7bc000       Deferred        sechost
PE             330260000-       330387000       Deferred        advapi32
PE             33ea00000-       33ea10000       Deferred        api-ms-win-crt-s
tring-l1-1-0
PE             344840000-       34484f000       Deferred        api-ms-win-crt-f
ilesystem-l1-1-0
PE             350a30000-       350a40000       Deferred        api-ms-win-crt-c
onvert-l1-1-0
PE             355100000-       35510f000       Deferred        api-ms-win-crt-l
ocale-l1-1-0
PE             360a80000-       360a91000       Deferred        api-ms-win-crt-m
ath-l1-1-0
PE             39b510000-       39b51f000       Deferred        api-ms-win-crt-h
eap-l1-1-0
PE             3af670000-       3af9dd000       Deferred        ucrtbase
PE             3afd00000-       3afd62000       Deferred        imm32
PE          7c6788a20000-    7c67892f2000       Deferred        shell32
Threads:
process  tid      prio (all id:s are in hex)
00000038 services.exe
        0000003c    0
        00000040    0
        00000054    0
        00000080    0
        00000098    0
        000000b0    0
        000000cc    0
        000000f4    0
        00000110    0
00000044 AppleMobileDeviceService.exe
        00000048    0
        0000005c    0
        00000064    0
        00000068    0
        0000006c    0
        000000e8    0
        000000ec    0
        000000f0    0
        00000158    0
0000004c explorer.exe
        00000050    0
        0000007c    0
        00000094    0
00000074 mDNSResponder.exe
        00000078    0
        00000084    0
        00000088    0
0000008c svchost.exe
        00000090    0
        0000009c    0
        000000a0    0
000000a4 winedevice.exe
        000000a8    0
        000000b8    0
        000000bc    0
        000000c0    0
000000c4 winedevice.exe
        000000c8    0
        000000d0    0
        000000d4    0
        000000d8    0
        000000dc    0
000000e0 atwtusb.exe
        000000e4    0
        000000f8    0
00000100 plugplay.exe
        00000104    0
        00000114    0
        00000118    0
        0000011c    0
        00000150    0
        00000154    0
00000108 atwtusb.exe
        0000010c    0
        0000014c    0
00000164 (D) Z:\home\spyro\.config\itch\apps\wavetracker\WaveTracker 1.0.1 (64-B
it)\WaveTracker.exe
        00000168    0 <==
        0000017c    0
0000016c conhost.exe
        00000170    0
System information:
    Wine build: wine-6.0.3 (Ubuntu 6.0.3~repack-1)
    Platform: x86_64
    Version: Windows 10
    Host system: Linux
    Host version: 6.8.0-40-generic

**Comment by 0hStormy** [comment]
> For me, it just crashes. It just doesn't show anything except an error in the terminal. (There should finally be a Linux version.) Here is the error:
> 
> wine WaveTracker.exe wine: Call from 000000007BC2ED58 to unimplemented function KERNEL32.dll.GetProcessGroupAffinity, aborting wine: Unhandled page fault on read access to 000000000000000C at address 0000000000990EF8 (thread 0168), starting debugger... Unhandled exception: page fault on read access to 0x0000000c in 64-bit code (0x0 000000000990ef8). Register dump: rip:0000000000990ef8 rsp:000000000029d250 rbp:000000000029f050 eflags:00010206 ( R- -- I - -P- ) rax:00000000000db340 rbx:000000000029d2a0 rcx:0000000000000130 rdx:000000000000 0000 rsi:000000000029ef20 rdi:000000000029dca0 r8:00000000000003c9 r9:000000001993 0520 r10:0000000000000005 r11:000000000029d5c0 r12:0000000000000000 r13:000000000029d3c8 r14:000000000029 fea0 r15:000000000029de08 Stack dump: 0x000000000029d250: 000000000029de08 000000000029fea0 0x000000000029d260: 0000000000a5f970 000000007b070687 0x000000000029d270: 000000000029f0f0 0000000000b52793 0x000000000029d280: 0000000000000004 000000000029d550 0x000000000029d290: 000000000029ef20 0000000000d89ef0 0x000000000029d2a0: 0000000000000000 0000000000a5e490 0x000000000029d2b0: 0000000000000000 0000000000a80257 0x000000000029d2c0: 0000000000000000 0000000000d89ef0 0x000000000029d2d0: 000000000029dca0 0000000000a5e3d1 0x000000000029d2e0: 000000000029d3c8 000000000029ef20 0x000000000029d2f0: 000000000029dca0 000000000029d550 0x000000000029d300: 000000000029ea20 0000000000a60980 Backtrace: =>0 0x0000000000990ef8 EntryPoint+0xffffffffffffffff() in coreclr (0x00000000002 9f050) 1 0x0000000000b52793 EntryPoint+0xf5c42() in coreclr (0x000000000029f050) 2 0x0000000000a80257 EntryPoint+0x23706() in coreclr (0x000000000029f050) 3 0x0000000000a60980 EntryPoint+0x3e2f() in coreclr (0x000000000029d550) 4 0x0000000000a5fa85 EntryPoint+0x2f34() in coreclr (0x000000000029d550) 5 0x000000007bc525a5 EntryPoint+0xffffffffffffffff() in ntdll (0x000000000029d 550) 6 0x00000000009f9e85 EntryPoint+0xffffffffffffffff() in coreclr (0x00000000002 9f3d1) 0x0000000000990ef8 EntryPoint+0xffffffffffffffff in coreclr: movl 0x000000 000000000c(%rdx),%eax Modules: Module Address Debug info Name (36 modules ) PE 890000- 8f4000 Deferred hostpolicy PE 900000- de9000 Export coreclr PE 7b000000- 7b3fd000 Deferred kernelbase PE 7b600000- 7b969000 Deferred kernel32 PE 7bc00000- 7bf3c000 Dwarf ntdll PE 140000000- 140045000 Deferred wavetracker PE 180000000- 180059000 Deferred hostfxr PE 1c8db0000- 1c9097000 Deferred msvcrt PE 1cd860000- 1cd86f000 Deferred api-ms-win-crt-u tility-l1-1-0 PE 1d97a0000- 1d97ae000 Deferred api-ms-win-core- fibers-l1-1-1 PE 21a7e0000- 21a9ce000 Deferred setupapi PE 231ae0000- 231d60000 Deferred rpcrt4 PE 23d820000- 23dfc2000 Deferred user32 PE 241850000- 24185e000 Deferred api-ms-win-crt-e nvironment-l1-1-0 PE 262250000- 262260000 Deferred api-ms-win-crt-r untime-l1-1-0 PE 26b4c0000- 26bbb9000 Deferred gdi32 PE 2739c0000- 273ec9000 Deferred oleaut32 PE 28ba60000- 28ba6e000 Deferred api-ms-win-crt-t ime-l1-1-0 PE 2e3540000- 2e364f000 Deferred shlwapi PE 2e8f10000- 2e948f000 Deferred ole32 PE 2f1fa0000- 2f1fbe000 Deferred version PE 30a2c0000- 30a2d0000 Deferred api-ms-win-crt-s tdio-l1-1-0 PE 30c980000- 30c98f000 Deferred api-ms-win-core- synch-l1-2-0 PE 3126f0000- 312744000 Deferred shcore PE 327020000- 3271a4000 Deferred combase PE 32a700000- 32a7bc000 Deferred sechost PE 330260000- 330387000 Deferred advapi32 PE 33ea00000- 33ea10000 Deferred api-ms-win-crt-s tring-l1-1-0 PE 344840000- 34484f000 Deferred api-ms-win-crt-f ilesystem-l1-1-0 PE 350a30000- 350a40000 Deferred api-ms-win-crt-c onvert-l1-1-0 PE 355100000- 35510f000 Deferred api-ms-win-crt-l ocale-l1-1-0 PE 360a80000- 360a91000 Deferred api-ms-win-crt-m ath-l1-1-0 PE 39b510000- 39b51f000 Deferred api-ms-win-crt-h eap-l1-1-0 PE 3af670000- 3af9dd000 Deferred ucrtbase PE 3afd00000- 3afd62000 Deferred imm32 PE 7c6788a20000- 7c67892f2000 Deferred shell32 Threads: process tid prio (all id:s are in hex) 00000038 services.exe 0000003c 0 00000040 0 00000054 0 00000080 0 00000098 0 000000b0 0 000000cc 0 000000f4 0 00000110 0 00000044 AppleMobileDeviceService.exe 00000048 0 0000005c 0 00000064 0 00000068 0 0000006c 0 000000e8 0 000000ec 0 000000f0 0 00000158 0 0000004c explorer.exe 00000050 0 0000007c 0 00000094 0 00000074 mDNSResponder.exe 00000078 0 00000084 0 00000088 0 0000008c svchost.exe 00000090 0 0000009c 0 000000a0 0 000000a4 winedevice.exe 000000a8 0 000000b8 0 000000bc 0 000000c0 0 000000c4 winedevice.exe 000000c8 0 000000d0 0 000000d4 0 000000d8 0 000000dc 0 000000e0 atwtusb.exe 000000e4 0 000000f8 0 00000100 plugplay.exe 00000104 0 00000114 0 00000118 0 0000011c 0 00000150 0 00000154 0 00000108 atwtusb.exe 0000010c 0 0000014c 0 00000164 (D) Z:\home\spyro.config\itch\apps\wavetracker\WaveTracker 1.0.1 (64-B it)\WaveTracker.exe 00000168 0 <== 0000017c 0 0000016c conhost.exe 00000170 0 System information: Wine build: wine-6.0.3 (Ubuntu 6.0.3~repack-1) Platform: x86_64 Version: Windows 10 Host system: Linux Host version: 6.8.0-40-generic

You seem to have a pretty old version of Wine (maybe old LTS) and I'm using Wine 9.15 which might have fixed issues in .Net

**Comment by Spyro24** [comment]
its wine 6.0 that ships with my distro. every windows programm works with that version. (except the tracker)

**Comment by 0hStormy** [comment]
Hold on, I'm gonna spin up an Ubuntu VM and test WaveTracker

**Comment by Spyro24** [comment]
Im using Mint 21.3 and i cant upgrade wine.

**Comment by Spyro24** [comment]
now im stuck with a black window :(

**Comment by 0hStormy** [comment]
> Im using Mint 21.3 and i cant upgrade wine.

You should probably upgrade to Mint 22. Also I seem to be having a black screen on Wine stable, trying Wine staging

**Comment by Spyro24** [comment]
I have to wait to upgrade to mint 22 (I dont wipe my system to upgrade.)

**Comment by 0hStormy** [comment]
I think it has to do with Wine stable vs staging. Wine stable doesn't seem to work, but staging does.

**Comment by Spyro24** [comment]
Staging shows me a black screen and proton will dont work! I HATE IT!!!

**Comment by 0hStormy** [comment]
Huh, weird. I don't really know how to help.

**Comment by Ferdi265** [comment]
I got it to work on Arch Linux with Wine 9.16 in a 64-bit prefix (`WINEARCH=win64`) using `winetricks dotnetdesktop8` and `winetricks dxvk`. It works really well like this, but without DXVK it crashed immediately. Maybe this helps some people where it doesn't work by default.

**Comment by 0hStormy** [comment]
> I got it to work on Arch Linux with Wine 9.16 in a 64-bit prefix (`WINEARCH=win64`) using `winetricks dotnetdesktop8` and `winetricks dxvk`. It works really well like this, but without DXVK it crashed immediately. Maybe this helps some people where it doesn't work by default.

Ah I just remember I had DXVK installed

**Comment by SolarLune** [comment]
@Ferdi265 - Perfect, that's it - I needed to install `dxvk` and dot net 8 (I installed it from the downloadable Windows executable, but I'm sure `dotnetdesktop8` would've worked too if it was the same file).

I'm now able to run it flawlessly on Linux, EndeavourOS. Thanks for the testing!

**Comment by huantianad** [comment]
> This is definitely something high on my priority list. Yes- really the only major things holding this back from being cross platform is the WASAPI audio (which I hope would be easy to swap out with any other audio API, all im doing is feeding it a buffer of samples... but I haven't explored this yet) and the Text/File dialogs. I'd really like in the future to have the custom-made TextBox and NumberBox to accept typing input, so that it matches the aesthetic of the other UI elements, provides better UX since you'd be able to type right in the box. Plus they'd be able to work cross-platform. I also might not be opposed to building custom file dialogs, after all I kind of did that with the sample browser, but I think using the platform's native dialog is the better move.
> 
> In the meantime, was anyone able to get it running in WINE? and if so what steps needed to be taken? If I could provide a WINE-compatible build that would give me a bit of peace of mind that Linux and MacOS users aren't _completely_ excluded.

@squiggythings Please use platform native file dialogues! Custom file dialogues, while aesthetic, are always missing features and don't integrate well into systems. Ideally for Linux, you should be using the FileChooser portal that's part of XDG Desktop Portals.

**Comment by darkhog** [comment]
Been a while since I've used Linux, but can't you just run it via Mono? Like this:

```
mono wavetracker.exe
```

**Comment by Speykious** [comment]
No. If that was the case we wouldn't need to write any code since you could just compile the project for Linux. But WaveTracker doesn't just depend on Monogame, it also depends on Windows-specific dependencies like WinForms for dialogs and NAudio for midi input and audio output, and it only supports Windows.

You can see some of the work needed for Linux support in my PR #28.

**Comment by V-Gira** [comment]
I wrote a quick guide on [how to use the Bottles' Flatpak to run the app on Linux](https://github.com/squiggythings/WaveTracker/discussions/45) for people having issues with the Wine version shipped in their Distro's repositories.

@Speykious and @squiggythings , hit me up if you want to package the app as a Flatpak (or even a Snap) and distribute it on Flathub once there's a Linux version. It's not something I've done but I've been meaning to figure it out and I'll be happy to help!

**Comment by Speykious** [comment]
Personally I'll probably make an AUR package since that's how I install my apps :v

**Comment by Averyfrog** [comment]
I'd be happy to attempt putting this on nixpkgs if thats ok!

**Comment by huantianad** [comment]
> I'd be happy to attempt putting this on nixpkgs if thats ok!

I haven't seen a windows app being run with wine packaged in nixpkgs, is that even possible?

**Comment by Speykious** [comment]
I think they mean once Linux support is merged in.

**Comment by ArcOfDream** [comment]
I think one of my main pain points with Wavetracker on Wine is that I'm totally unable to import any samples into the editor. Tried both with Bottles and outside of Bottles in a separate prefix.

Haven't been really able to tell if it's because I have Pipewire set up on the PC or something else.

**Comment by 0hStormy** [comment]
> I think one of my main pain points with Wavetracker on Wine is that I'm totally unable to import any samples into the editor. Tried both with Bottles and outside of Bottles in a separate prefix. Drag and dropping them in also seems to kill the program entirely.
> 
> 
> 
> Haven't been really able to tell if it's because I have Pipewire set up on the PC or something else.

90% chance you have pipewire, also you're not alone with not being able to import samples, I can't either.

**Comment by desttinghim** [comment]
After downloading and building the latest commit of WaveTracker (commit starting with 741575cfe) on Arch Linux, WaveTracker was not playing sound for me. WaveTracker kept printing this message while running: 
```
ALSA lib pcm_dmix.c:1000:(snd_pcm_dmix_open) unable to open slave
```
When I attempted to change the audio device and save the changes, WaveTracker would hang and need to be killed using Ctrl+C in the terminal.

The good news is that the fix was simple! I simply needed to install `pipewire-alsa` using pacman. I'm posting this here in case anyone else has this problem, so they don't need to search around as much for the answer.

**Comment by Speykious** [comment]
> I simply needed to install `pipewire-alsa` using pacman.

@user0-07161 you might want to add the dependency to the AUR package
Edit: I'm now co-maintainer of the AUR package, I updated it to have`pipewire-alsa` and `pulseaudio-alsa` as `optdepends`
